helix-core/src/main/java/org/apache/helix/controller/stages/ManagementModeStage.java

```diff
@@ -36,13 +38,30 @@
   @Override
   public void process(ClusterEvent event) throws Exception {
     // TODO: implement the stage
+    _eventId = event.getEventId();
     String clusterName = event.getClusterName();
+    ClusterStatusMonitor clusterStatusMonitor =
+        event.getAttribute(AttributeName.clusterStatusMonitor.name());
     ManagementControllerDataProvider cache =
         event.getAttribute(AttributeName.ControllerDataProvider.name());
-    if (!HelixUtil.inManagementMode(cache)) {
-      LOG.info("Exiting management mode pipeline for cluster {}", clusterName);
+
+    checkInManagementMode(clusterStatusMonitor, clusterName, cache);
+  }
+
+  private void checkInManagementMode(ClusterStatusMonitor clusterStatusMonitor,
+      String clusterName, ManagementControllerDataProvider cache) throws StageException {
+    // Should exit management mode
+    if (!HelixUtil.inManagementMode(cache.getPauseSignal(), cache.getLiveInstances(),
```

Comment:
As we discussed, pending message of freeze check is necessary to protect whether we can break this pipeline. Otherwise, there would be a race condition can cause lot of unnecessary traffic to ZK.